libraries/AP_HAL_Linux/AnalogIn_Navio.cpp

```diff
@@ -0,0 +1,147 @@
+#include <AP_HAL/AP_HAL.h>
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX
+
+#include "AnalogIn_Navio.h"
+
+#include <cstdlib>
+#include <unistd.h>
+#include <cstdio>
+#include <errno.h>
+
+static const AP_HAL::HAL &hal = AP_HAL::get_HAL();
+
+union adc_params {
+    char channel[sizeof("XXXX")];
+};
+
+#define ADC_BASE_PATH "/sys/kernel/rcio/adc"
+#define ADC_PATH_MAX (sizeof(ADC_BASE_PATH) + sizeof(adc_params) - 1)
+
+void NavioAnalogSource::set_channel(uint8_t pin)
+{
+    char *channel_path;
+
+    if (pin == ANALOG_INPUT_NONE) {
+        return;
+    }
+
+    if (asprintf(&channel_path, "%s/ch%d", ADC_BASE_PATH, pin) == -1) {
+        AP_HAL::panic("asprintf failed\n");
+    }
+
+    if (_fd >= 0) {
+        ::close(_fd);
+    }
+
+    _fd = ::open(channel_path, O_RDONLY);
+
+    if (_fd < 0) {
+        hal.console->printf("%s not opened: %s\n", channel_path, strerror(errno));
+    }
+
+    free(channel_path);
+}
+
+NavioAnalogSource::NavioAnalogSource(uint8_t pin):
+    _pin(pin)
+{
+    set_channel(pin);
+}
+
+void NavioAnalogSource::set_pin(uint8_t pin)
+{
+    if (_pin == pin) {
+        return;
+    }
+
+    set_channel(pin);
+
+    _pin = pin;
+}
+
+float NavioAnalogSource::read_average()
+{
+    return read_latest();
+}
+
+float NavioAnalogSource::read_latest()
+{
+    return _value;
+}
+
+float NavioAnalogSource::voltage_average()
+{
+    char buffer[ADC_PATH_MAX];
+
+    if (pread(_fd, buffer, sizeof(buffer), 0) < 0) {
```

Comment:
it's actually reading their ADC driver, /sys/kernel/rcio/adc. It may not make much sense to return 0.  Checking it doesn't hurt neither so this should be changed to `<= 0`.

`buffer` in this case would contain whatever was on the stack.